core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+      this.currentVersionId = newVersionId;
 
-    if (versionHistorySizeToKeep <= 0) {
-      LOG.warn(
-          "{} must be positive but was {}",
+      if (lastAddedVersionId != null && lastAddedVersionId == newVersionId) {
+        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(LAST_ADDED));
+      } else {
+        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+      }
+
+      return this;
+    }
+
+    private int highestFieldId() {
+      return schemas.stream().map(Schema::highestFieldId).max(Integer::compareTo).orElse(0);
+    }
+
+    public Builder setCurrentVersion(ViewVersion version) {
+      Schema schema = indexSchemas(schemas).get(version.schemaId());
+      int newSchemaId = addSchemaInternal(schema);
+      return setCurrentVersionId(
+          addVersionInternal(
+              ImmutableViewVersion.builder().from(version).schemaId(newSchemaId).build()));
+    }
+
+    private int reuseOrCreateNewViewVersionId(ViewVersion viewVersion) {
+      // if the view version already exists, use its id; otherwise use the highest id + 1
+      int newVersionId = viewVersion.versionId();
+      for (ViewVersion version : versions) {
+        if (version.equals(viewVersion)) {
+          return version.versionId();
+        } else if (version.versionId() >= newVersionId) {
+          newVersionId = viewVersion.versionId() + 1;
+        }
+      }
+
+      return newVersionId;
+    }
+
+    private int addVersionInternal(ViewVersion version) {
+      int newVersionId = reuseOrCreateNewViewVersionId(version);
+
+      if (versions.stream().anyMatch(v -> v.versionId() == newVersionId)) {
+        boolean isNewVersion =
+            lastAddedVersionId != null
+                && changes(MetadataUpdate.AddViewVersion.class)
+                    .anyMatch(added -> added.viewVersion().versionId() == newVersionId);
+        this.lastAddedVersionId = isNewVersion ? newVersionId : null;
+        return newVersionId;
+      }
+
+      ViewVersion newVersion;
+      if (newVersionId != version.versionId()) {
+        newVersion = ImmutableViewVersion.builder().from(version).versionId(newVersionId).build();
+      } else {
+        newVersion = version;
+      }
+
+      this.versions.add(newVersion);
+      this.changes.add(new MetadataUpdate.AddViewVersion(newVersion));
+      this.history.add(
+          ImmutableViewHistoryEntry.builder()
+              .timestampMillis(newVersion.timestampMillis())
+              .versionId(newVersion.versionId())
+              .build());
+      this.lastAddedVersionId = newVersionId;
+
+      return newVersionId;
+    }
+
+    private int addSchemaInternal(Schema schema) {
+      if (schema.schemaId() == -1) {
+        ValidationException.check(
+            lastAddedSchemaId != null, "Cannot set last added schema: no schema has been added");
+
+        return addSchemaInternal(
+            new Schema(lastAddedSchemaId, schema.columns(), schema.identifierFieldIds()));
```

Comment:
If there is no last added schema ID -- which is set by adding a schema -- then this will fail with a `NullPointerException` that has no context. Because this is set by adding a schema, the only time this would succeed is when you've added a schema with a "normal" ID (e.g. 5) and then add a schema with ID -1 for some reason.

It gets weirder from there because this would use the last added schema ID to add a new schema with possibly different fields. Luckily, the schema to add would be sent through `reuseOrCreateNewSchemaId`.

As I noted above, I don't think it is valid to have a schema with `schemaId=-1` and to handle it differently. The correct behavior is to add the schema and control schema ID assignment in this builder, which is already done by `reuseOrCreateNewSchemaId`.

Reply:
should be fixed now with your changes